06_puppy/jimbotech/mapStore_test.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+type storesSuite struct {
+	suite.Suite
+	store  Storer
+	mapper mapCheck
+}
+
+func TestStorer(t *testing.T) {
+	m := NewPuppyStorer()
+	ms := &SyncMapStore{}
+	suite.Run(t, &storesSuite{store: m, mapper: m})
+	suite.Run(t, &storesSuite{store: ms, mapper: ms})
+}
+
+// TestMapStoreWithoutContructor may not be required as it may not be
+// possible to create an instance of that type without using the constructor
+func TestMapStoreWithoutContructor(t *testing.T) {
+	var puppyStore MapStore
+	pup := Puppy{1, "kelpie", "brown", "indispensable"}
+
+	_, err := puppyStore.CreatePuppy(&pup)
+	f := func(t *testing.T) {
+		assert.NotNil(t, err)
+		assert.Equal(t, "store not created, call initializing constructor first", err.Error())
+	}
+	t.Run("CreatePuppy test", f)
+
+	err = puppyStore.UpdatePuppy(1, &pup)
+	t.Run("UpdatePuppy test", f)
+	_, err = puppyStore.ReadPuppy(1)
+	t.Run("ReadPuppy test", f)
+	puppyStore.DeletePuppy(1)
+}
+
+// TestCreateSuccess add to the store and verify
+// by reading that it is in the store and then remove it
+// to leave the store in the original state
+func (s *storesSuite) TestCreateSuccess() {
+	asserter := assert.New(s.T())
+	if success, pup := create(s); success {
```

Comment:
I like to use `testify/require` for this - require makes the test fail right away, so I use it to verify that the test is set up properly and not continue if not.
```
success, pup := create(s)
require.True(success)
```

... now I've looked at `create()` I see that you probably don't want to use assert/require here. If you use `require` in `create()` you can assume it always returns a puppy, or does not return at all.

Reply:
Because I did not see this comment first, I spent a lot of time left aligning my code. Now with this comment I see that I wasted all that effort as this more or less left aligns it anyway. I've gone crazy with require now. Originally I assumed that is how assert works, as assert in other languages terminates the test. It's require for me all the way now ... and I thought I was so clever for testing the response from assert to avoid continuing ... kill me now, will this project ever end?